File: kits/kotlin/simple/bot/src/main/kotlin/luxaibot/lux/Unit.kt
package luxaibot.lux

class Unit(
    teamid: Int,
    u_type: Int,
    unitid: String,
    x: Int,
    y: Int,
    cooldown: Double,
    wood: Int,
    coal: Int,
    uranium: Int
) {
    var pos: Position
    var team: Int
    var id: String
    var type: Int
    var cooldown: Double
    var cargo: Cargo
    val isWorker: Boolean
        get() = type == 0
    val isCart: Boolean
        get() = type == 1
    val cargoSpaceLeft: Int
        get() {
            val spaceused: Int = cargo.wood + cargo.coal + cargo.uranium
            return if (type == GameConstants.UNIT_TYPES.WORKER) {
                GameConstants.PARAMETERS.RESOURCE_CAPACITY.WORKER - spaceused
            } else {
                GameConstants.PARAMETERS.RESOURCE_CAPACITY.CART - spaceused
            }
        }

    fun canBuild(gameMap: GameMap): Boolean {
        val cell = gameMap.getCellByPos(pos)
        return if (!cell!!.hasResource() && canAct() && cargo.wood + cargo.coal + cargo.uranium >= GameConstants.PARAMETERS.CITY_BUILD_COST) true else false
    }

    fun canAct(): Boolean {
        return cooldown < 1
    }

    fun move(dir: Direction?): String {
        return java.lang.String.format("m %s %s", id, dir)
    }

    fun transfer(destId: String?, resourceType: String?, amount: Int): String {
        return String.format("t %s %s %s %d", id, destId, resourceType, amount)
    }

    fun buildCity(): String {
        return String.format("bcity %s", id)
    }

    fun pillage(): String {
        return String.format("p %s", id)
    }

    init {
        pos = Position(x, y)
        team = teamid
        id = unitid
        type = u_type
        this.cooldown = cooldown
        cargo = Cargo(wood, coal, uranium)
    }
}

File: kits/csharp/simple/Bot/Lux/Unit.cs
namespace Bot.Lux
{
	public class Unit
	{
		public Position Pos;
		public int Team;
		public string Id;
		public int Type;
		public double Cooldown;
		public Cargo Cargo;

		public Unit(int teamId, int type, string unitId, int x, int y, double cooldown, int wood, int coal, int uranium)
		{
			Pos = new Position(x, y);
			Team = teamId;
			Id = unitId;
			Type = type;
			Cooldown = cooldown;
			Cargo = new Cargo(wood, coal, uranium);
		}

		public bool IsWorker()
		{
			return Type == 0;
		}

		public bool IsCart()
		{
			return Type == 1;
		}

		public int GetCargoSpaceLeft()
		{
			var spaceUsed = Cargo.Wood + Cargo.Coal + Cargo.Uranium;
			if (Type == GameConstants.UNIT_TYPES.WORKER)
			{
				return GameConstants.PARAMETERS.RESOURCE_CAPACITY.WORKER - spaceUsed;
			}

			return GameConstants.PARAMETERS.RESOURCE_CAPACITY.CART - spaceUsed;
		}

		public bool CanBuild(GameMap gameMap)
		{
			var cell = gameMap.GetCellByPos(Pos);
			return !cell.HasResource() && CanAct() && 
			       Cargo.Wood + Cargo.Coal + Cargo.Uranium >= GameConstants.PARAMETERS.CITY_BUILD_COST;
		}

		public bool CanAct()
		{
			return Cooldown < 1;
		}

		public string Move(string dir)
		{
			return $"m {Id} {dir}";
		}

		public string Transfer(string destId, string resourceType, int amount)
		{
			return $"t {Id} {destId} {resourceType} {amount}";
		}

		public string BuildCity()
		{
			return $"bcity {Id}";
		}

		public string Pillage()
		{
			return $"p {Id}";
		}
	}
}
